clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java

```diff
@@ -59,29 +63,67 @@ public PartitionData(long offset, int maxBytes) {
+        }
+
+        public static final List<TopicAndPartitionData> groupByTopicOrdered(Map<TopicPartition, PartitionData> fetchData) {
+            List<TopicAndPartitionData> topics = new ArrayList<TopicAndPartitionData>();
+            for (Map.Entry<TopicPartition, PartitionData> topicEntry : fetchData.entrySet()) {
+                String topic = topicEntry.getKey().topic();
+                int parition = topicEntry.getKey().partition();
+                PartitionData partitionData = topicEntry.getValue();
+                if (topics.isEmpty() || topics.get(topics.size() - 1).topic != topic) {
+                    topics.add(new TopicAndPartitionData(topic));
+                }
+                topics.get(topics.size() - 1).partitions.put(parition, partitionData);
+            }
+            return topics;
+        }
+    }
+
     /**
      * Create a non-replica fetch request
      */
     public FetchRequest(int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData) {
-        this(CONSUMER_REPLICA_ID, maxWait, minBytes, fetchData);
+        this(CONSUMER_REPLICA_ID, maxWait, minBytes, fetchData, ProtoUtils.latestVersion(ApiKeys.FETCH.id), DEFAULT_RESPONSE_MAX_BYTES);
     }
+    public FetchRequest(int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData, int maxBytes) {
+        this(CONSUMER_REPLICA_ID, maxWait, minBytes, fetchData, ProtoUtils.latestVersion(ApiKeys.FETCH.id), maxBytes);
+    }
+
 
     /**
      * Create a replica fetch request
      */
     public FetchRequest(int replicaId, int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData) {
-        super(new Struct(CURRENT_SCHEMA));
-        Map<String, Map<Integer, PartitionData>> topicsData = CollectionUtils.groupDataByTopic(fetchData);
+        this(replicaId, maxWait, minBytes, fetchData, ProtoUtils.latestVersion(ApiKeys.FETCH.id), DEFAULT_RESPONSE_MAX_BYTES);
+    }
+
+    public FetchRequest(int replicaId, int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData, int maxBytes) {
+        this(replicaId, maxWait, minBytes, fetchData, ProtoUtils.latestVersion(ApiKeys.FETCH.id), maxBytes);
+    }
+
+    public FetchRequest(int replicaId, int maxWait, int minBytes, Map<TopicPartition, PartitionData> fetchData, int version, int maxBytes) {
+        super(new Struct(ProtoUtils.requestSchema(ApiKeys.FETCH.id, version)));
+        List<TopicAndPartitionData> topicsData = TopicAndPartitionData.groupByTopicOrdered(fetchData);
 
         struct.set(REPLICA_ID_KEY_NAME, replicaId);
         struct.set(MAX_WAIT_KEY_NAME, maxWait);
         struct.set(MIN_BYTES_KEY_NAME, minBytes);
+        if (version >= 3)
+            struct.set(RESPONSE_MAX_BYTES_KEY_NAME, maxBytes);
         List<Struct> topicArray = new ArrayList<Struct>();
-        for (Map.Entry<String, Map<Integer, PartitionData>> topicEntry : topicsData.entrySet()) {
+        for (TopicAndPartitionData topicEntry : topicsData) {
```

Comment:
The fact that the partitions are grouped by topic means that we don't have full flexibility when implementing the round-robin fetch. Have you thought about how this will work exactly?